### PR TITLE
fix: zkSync deployments contract addresses instead of canonical by default

### PR DESCRIPTION
**Pull request by vmosharova**
> zkSync asks for zksync,
> canonical bytecode leaks through—
> guard on registration.

## What it solves

Resolves: [WA-2000](https://linear.app/safe-global/issue/WA-2000/multisend-and-simulations-not-working-on-zksync-lens-chains)
The "first network address" fallback handed back an address that was never deployed on the target chain. `getChainAgnosticAddress(..., 'zksync')` could return the zkSync-type address on chains where no zkSync deployment is registered, or return the canonical (EVM) address.

## How this PR fixes it

When a non-default `deploymentType` (e.g. `'zksync'`, `'eip155'`) is requested, `getChainAgnosticAddress` now returns the type-specific address only if it is explicitly listed under `networkAddresses[chainId]`. If not registered, it falls through to the existing per-chain and chain-agnostic lookups. So we do not return an address that was never deployed on the target chain.

## How to test it

1. Run `yarn workspace @safe-global/utils test packages/utils/src/services/contracts/__tests__/deployments.test.ts`
2. Verify the new `eip155 MultiSendCallOnly v1.3.0` and `zkSync Era MultiSendCallOnly v1.3.0 / v1.4.1` describe blocks all pass, including the `does NOT return ...` negative guards
3. Manually: on zkSync Era (chain 324), build a transaction that resolves the MultiSendCallOnly address via `resolveChainAgnosticContractAddresses(version, true, true)` and confirm it returns the zkSync-specific EraVM address (not the canonical EVM one)

## Visual summary

```mermaid
flowchart TB
  subgraph Before_broken
    A1["getChainAgnosticAddress(dep, '324', 'zksync')"] --> B1["getCanonicalOrFirstAddress(dep, '324')"]
    B1 --> C1["networkAddresses['324'] = ['zksync','canonical','eip155']"]
    C1 --> D1["hasCanonical? yes → return canonical"]
    D1 --> E1["❌ EVM address, EraVM Safe cannot delegatecall"]
  end
  subgraph After_fixed
    A2["getChainAgnosticAddress(dep, '324', 'zksync')"] --> B2["typeAddress = deployments.zksync.address"]
    B2 --> C2["Is typeAddress in networkAddresses['324']?"]
    C2 -->|Yes| D2["✅ return zksync EraVM address"]
    C2 -->|No, e.g. chain 1| E2["fall through to canonical-or-first"]
  end
```

## Checklist

- [ ] I've tested the branch on mobile 📱 N/A
- [ ] I've documented how it affects the analytics (if at all) 📊 N/A
- [x] I've written a unit/e2e test for it (if applicable) 🧑‍💻

---

## CLA signature

With the submission of this Pull Request, I confirm that I have read and agree to the terms of the [Contributor License Agreement](https://safe.global/cla).

### CODE REVIEW

**Comment by Zalens** [comment]
I think this is a good fix to keep the purpose of my former PR and fix for specific chains. Well done!

**Comment by katspaugh** [comment]
@claude review

File: packages/utils/src/services/contracts/deployments.ts [hunk 1]
```diff
@@ -264,14 +280,38 @@ export const isCanonicalDeployment = (
...
-  const safeVersion = version ?? '1.3.0'
-  const deployment = getMultiSendCallOnlyDeployments({ version: safeVersion })
-  return deployment?.deployments.canonical?.address
+export const getCanonicalMultiSendCallOnlyAddress = (
+  chainId: string,
+  version: SafeState['version'],
+): string | undefined => {
+  const requested = (version ?? '1.3.0') as CanonicalFallbackVersion
+  const startIndex = CANONICAL_FALLBACK_VERSIONS.indexOf(requested)
+  const candidates = startIndex >= 0 ? CANONICAL_FALLBACK_VERSIONS.slice(startIndex) : CANONICAL_FALLBACK_VERSIONS
```
[the remark this answers is not in the document]
Comment (vmosharova) [reply packages/utils/src/services/contracts/deployments.ts]: Addressed in [9acf000](https://github.com/safe-global/safe-wallet-monorepo/pull/7704/commits/9acf00025ee40429ca73614764194c11778cbedb)

File: packages/utils/src/services/contracts/deployments.ts [hunk 2]
```diff
@@ -264,14 +280,38 @@ export const isCanonicalDeployment = (
...
 
+// Newest --> oldest (fallback candidates) starting from the requested version
+const CANONICAL_FALLBACK_VERSIONS = ['1.5.0', '1.4.1', '1.3.0'] as const
+type CanonicalFallbackVersion = (typeof CANONICAL_FALLBACK_VERSIONS)[number]
+
+const isCanonicalRegisteredOnChain = (deployment: SingletonDeploymentV2 | undefined, chainId: string): boolean => {
+  const canonicalAddress = deployment?.deployments.canonical?.address
+  if (!canonicalAddress) return false
+  const networkAddresses = toNetworkAddressList(deployment.networkAddresses[chainId] ?? [])
+  return networkAddresses.some((addr) => sameAddress(addr, canonicalAddress))
```
[the remark this answers is not in the document]
Comment (vmosharova) [reply packages/utils/src/services/contracts/deployments.ts]: Addressed in [9acf000](https://github.com/safe-global/safe-wallet-monorepo/pull/7704/commits/9acf00025ee40429ca73614764194c11778cbedb)

File: packages/utils/src/services/contracts/__tests__/deployments.test.ts [hunk 3]
```diff
@@ -11,6 +12,12 @@ import {
   resolveChainAgnosticContractAddresses,
 } from '../../contracts/deployments'
 import { ZKSYNC_ERA_CHAIN_ID } from '../../../config/chains'
+import type { SafeState } from '@safe-global/store/gateway/AUTO_GENERATED/safes'
+
+// picks the zk address for 1.5.0 if/when it has a zkSync deployment:
+const ZKSYNC_MULTISEND_CALL_ONLY_VERSIONS = (['1.3.0', '1.4.1', '1.5.0'] as const).filter(
+  (v) => getMultiSendCallOnlyDeployments({ version: v })?.deployments.zksync?.address,
+)
```
[the remark this answers is not in the document]
Comment (vmosharova) [reply packages/utils/src/services/contracts/__tests__/deployments.test.ts:21]: Addressed in [9acf000](https://github.com/safe-global/safe-wallet-monorepo/pull/7704/commits/9acf00025ee40429ca73614764194c11778cbedb)

**Review by katspaugh — APPROVED** [review]

**Comment by katspaugh** [comment]
## PR Briefing

[Watch the PR walkthrough](https://github.com/5afe/wallet-pr-artifacts/releases/download/prbrief-safe-global-safe-wallet-monorepo-pr-7704-9acf00025ee4/video-github.mp4)

▶ **[Watch the walkthrough video](https://github.com/5afe/wallet-pr-artifacts/releases/download/prbrief-safe-global-safe-wallet-monorepo-pr-7704-9acf00025ee4/video-github.mp4)** (opens MP4 in a new tab)

This PR fixes incorrect MultiSendCallOnly contract address resolution on zkSync Era and similar chains by updating `getChainAgnosticAddress` in `packages/utils/src/services/contracts/deployments.ts` to only return a type-specific address (e.g. `'zksync'`) when that address is explicitly registered under `networkAddresses[chainId]`, rather than falling back to a canonical EVM address that was never deployed on the target chain. Additionally, `getCanonicalMultiSendCallOnlyAddress` gains a `chainId` parameter and now iterates a `CANONICAL_FALLBACK_VERSIONS` list to find the newest canonical address actually registered on the given chain. Both `apps/web` and `apps/mobile` call sites of `getCanonicalMultiSendCallOnlyAddress` are updated to pass `chainId` as the first argument. A comprehensive test suite in `deployments.test.ts` validates the fix with parameterized cases covering zkSync, eip155, canonical, and fallback scenarios.

<details>
<summary><b>Full briefing</b> (click to expand)</summary>

## Summary

This PR fixes incorrect MultiSendCallOnly contract address resolution on zkSync Era and similar chains by updating `getChainAgnosticAddress` in `packages/utils/src/services/contracts/deployments.ts` to only return a type-specific address (e.g. `'zksync'`) when that address is explicitly registered under `networkAddresses[chainId]`, rather than falling back to a canonical EVM address that was never deployed on the target chain. Additionally, `getCanonicalMultiSendCallOnlyAddress` gains a `chainId` parameter and now iterates a `CANONICAL_FALLBACK_VERSIONS` list to find the newest canonical address actually registered on the given chain. Both `apps/web` and `apps/mobile` call sites of `getCanonicalMultiSendCallOnlyAddress` are updated to pass `chainId` as the first argument. A comprehensive test suite in `deployments.test.ts` validates the fix with parameterized cases covering zkSync, eip155, canonical, and fallback scenarios.

## Architecture notes

The fix is concentrated in `packages/utils/src/services/contracts/deployments.ts`, which is shared by both the web and mobile apps. The data flow change is: `getChainAgnosticAddress` now checks whether the type-specific address appears in `networkAddresses[chainId]` before returning it, preventing canonical EVM addresses from leaking into zkSync deployments. `getCanonicalMultiSendCallOnlyAddress` now accepts `chainId` and uses `hasCanonicalDeployment` with a hardcoded `CANONICAL_FALLBACK_VERSIONS` array (`['1.5.0', '1.4.1', '1.3.0']`), meaning future Safe versions must be manually added to this list. Both app-layer call sites (`apps/web/src/services/contracts/safeContracts.ts` and `apps/mobile/src/services/contracts/safeContracts.ts`) are updated to pass `chainId`, so any missed call site would silently break address resolution.

## Ripple risks

Analyzed 5 changed exports. No untouched external callers found — all usages appear to be inside the PR diff.

_Analyzed 4 file(s), 5 changed export(s), 0 uncovered callsite(s) in 2.0s._

## Walkthrough outline

_18 steps across 5 chapters, approximately 4 min 15s._

### Why this broke in the first place

1. **motivation** · `packages/utils/src/services/contracts/deployments.ts` L116-130 `getChainAgnosticAddress`
   Here is ground zero. When you asked for a zkSync-type MultiSendCallOnly address, the old code would fall straight through to the canonical EVM address if no type-specific match was found. On zkSync Era, that canonical address was never deployed, so any delegatecall would silently fail.

2. **motivation** · `packages/utils/src/services/contracts/deployments.ts` L116-145
   This diagram captures the before-and-after data flow so the fix is crystal clear before we dive into code.

### The core fix in deployments.ts

3. **change** · `packages/utils/src/services/contracts/deployments.ts` L73-85 `hasCanonicalDeployment`
   First stop in the fix: `hasCanonicalDeployment`. It now accepts a `chainId` and checks whether the canonical address actually appears in that chain's `networkAddresses` list. This is the predicate that powers the new canonical fallback loop.

4. **change** · `packages/utils/src/services/contracts/deployments.ts` L116-135 `getChainAgnosticAddress`
   Now the guard itself. When `deploymentType` is not `'canonical'`, the function grabs the type-specific address and checks whether it lives in `networkAddresses[chainId]`. If it does not, we fall through — no more leaking EVM addresses onto EraVM chains.

5. **change** · `packages/utils/src/services/contracts/deployments.ts` L135-150 `getChainAgnosticAddress`
   The fallthrough path is unchanged — `getCanonicalOrFirstAddress` still handles chains that have no type-specific deployment. So non-zkSync chains keep working exactly as before.

6. **change** · `packages/utils/src/services/contracts/deployments.ts` L290-320 `getCanonicalMultiSendCallOnlyAddress`
   `getCanonicalMultiSendCallOnlyAddress` is the other big change. It now takes `chainId` as its first argument and iterates a hardcoded `CANONICAL_FALLBACK_VERSIONS` list — newest first — returning the first version that `hasCanonicalDeployment` confirms is actually registered on that chain.

7. **change** · `packages/utils/src/services/contracts/deployments.ts` L280-292
   Here is `CANONICAL_FALLBACK_VERSIONS`, hardcoded as `['1.5.0', '1.4.1', '1.3.0']`. The order matters — we want the newest canonical address available on the chain. Any future Safe version must be manually prepended to this list.

8. **change** · `packages/utils/src/services/contracts/deployments.ts` L315-325 `getCanonicalMultiSendCallOnlyAddress`
   If no version in the fallback list is registered on the chain, the function emits a `console.warn` and returns `undefined`. Every caller must now handle that `undefined` gracefully — we will check the call sites next.

9. **change** · `packages/utils/src/services/contracts/__tests__/deployments.test.ts` L1-40
   The test file is heavily expanded. Three new describe blocks cover zkSync, eip155, and canonical scenarios with parameterized cases. There is also a guard test that fails loudly if the `ZKSYNC_MULTISEND_CALL_ONLY_VERSIONS` list ever becomes empty — preventing silent zero-test suites.

### Callers that had to adapt

10. **ripple** · `apps/web/src/services/contracts/safeContracts.ts` L83-97 `getReadOnlyMultiSendCallOnlyContract`
   Web app call site. `getReadOnlyMultiSendCallOnlyContract` now passes `chainId` as the first argument to `getCanonicalMultiSendCallOnlyAddress`. The `chainId` is already in scope here from the surrounding guard, so no new plumbing was needed.

11. **ripple** · `apps/mobile/src/services/contracts/safeContracts.ts` L27-38 `getReadOnlyMultiSendCallOnlyContract`
   Same change on the mobile side. `chainId` is threaded in as the first argument. Both apps now agree on the calling convention — any missed call site would silently pass `undefined` as `chainId` and likely return `undefined` from the lookup.

12. **ripple** · `packages/utils/src/services/contracts/__tests__/deployments.test.ts` L80-120
   The rewritten `getCanonicalMultiSendCallOnlyAddress` describe block validates the fallback chain. One test confirms it falls from v1.5.0 to v1.4.1 on zkSync chain 324, and another confirms it returns `undefined` and warns when no canonical version is registered.

13. **ripple** · `packages/utils/src/services/contracts/__tests__/deployments.test.ts` L150-200
   The zkSync and eip155 describe blocks use `describe.each` over a dynamically computed version list. The negative guard tests — labeled `does NOT return` — are the most important: they confirm the old leakage path is closed.

### What could still go wrong

14. **risk** · `packages/utils/src/services/contracts/deployments.ts` L280-292
   `CANONICAL_FALLBACK_VERSIONS` is a manually maintained list. When Safe v1.6.0 ships, someone must prepend it here or the new version will never be returned as a canonical address — there is no automated enforcement of this.

15. **risk** · `packages/utils/src/services/contracts/deployments.ts` L315-325 `getCanonicalMultiSendCallOnlyAddress`
   When the fallback loop exhausts all versions, it returns `undefined`. Both app call sites need to handle that — if either passes `undefined` downstream as a contract address, you get a silent runtime failure rather than a clear error.

16. **risk** · `packages/utils/src/services/contracts/__tests__/deployments.test.ts` L40-65
   The test chain ID constants — `SOPHON='50104'`, `CANONICAL_ONLY_CHAIN_ID='30'`, and friends — are hardcoded assumptions about the current `safe-deployments` package data. If that upstream package changes its network registrations, these tests could silently validate wrong invariants.

17. **risk** · `packages/utils/src/services/contracts/deployments.ts` L116-125 `getChainAgnosticAddress`
   The new guard fires when `deploymentType !== 'canonical'`, which correctly catches both `'zksync'` and `'eip155'`. Worth a quick read to confirm no other deployment type strings exist in the wild that might slip through unguarded.

### The bottom line

18. **wrap** · `packages/utils/src/services/contracts/deployments.ts` L116-145 `getChainAgnosticAddress`
   The bottom line: one registration check in `getChainAgnosticAddress` closes the canonical-address leak, `getCanonicalMultiSendCallOnlyAddress` now version-walks per chain, and both apps pass `chainId` through — MultiSend on zkSync Era finally gets the EraVM address it was always supposed to have.

## Reviewer checklist

- [ ] Verify all call sites of `getCanonicalMultiSendCallOnlyAddress` across both `apps/web` and `apps/mobile` have been updated to pass `chainId` as the first argument, and that `chainId` is never undefined or empty string at those call sites.
- [ ] Confirm that `getChainAgnosticAddress` correctly handles the `'eip155'` deployment type (not just `'zksync'`) under the new guard, and that the fallback to `getCanonicalOrFirstAddress` behaves correctly for chains with no registered type-specific address.
- [ ] Check that callers of `getCanonicalMultiSendCallOnlyAddress` handle the `undefined` return value gracefully when no canonical deployment is found for any version in `CANONICAL_FALLBACK_VERSIONS` on the given chain.
- [ ] Validate that the hardcoded chain ID constants used in the new tests (e.g. SOPHON='50104', EIP155_ONLY_V130='16', CANONICAL_ONLY_CHAIN_ID='30') still accurately reflect the current `safe-deployments` package registrations, and that the `ZKSYNC_MULTISEND_CALL_ONLY_VERSIONS` guard test runs before the `describe.each` blocks to prevent silent empty suites.
- [ ] Ensure the `toNetworkAddressList` helper safely handles chains with no `networkAddresses` entry (the `?? []` fallback) and that `hasCanonicalDeployment` returns false rather than throwing when passed an empty list.

## File-by-file

### `apps/mobile/src/services/contracts/safeContracts.ts`

In `getReadOnlyMultiSendCallOnlyContract`, the call to `getCanonicalMultiSendCallOnlyAddress` is updated to pass `chainId` as the first argument, changing the signature from `(safeVersion)` to `(chainId, safeVersion)`.

_Passing `chainId` allows `getCanonicalMultiSendCallOnlyAddress` to return the chain-specific (e.g. zkSync EraVM) canonical address rather than a chain-agnostic fallback, fixing incorrect address resolution on zkSync and similar chains._

**Risks:**
- Verify the updated signature of `getCanonicalMultiSendCallOnlyAddress` actually accepts `chainId` as first param and that all other call sites have been updated consistently
- Check that passing `chainId` does not break behavior on non-zkSync chains where the canonical address should remain unchanged
- Confirm `isCanonicalDeployment` and `getCanonicalMultiSendCallOnlyAddress` use the same chainId format/type to avoid subtle mismatches

### `apps/web/src/services/contracts/safeContracts.ts`

In `getReadOnlyMultiSendCallOnlyContract`, the call to `getCanonicalMultiSendCallOnlyAddress` is updated to pass `chainId` as the first argument (previously called with only `version`).

_Passing `chainId` allows `getCanonicalMultiSendCallOnlyAddress` to perform chain-specific address lookup, ensuring zkSync and other non-EVM chains return the correct deployed address rather than the canonical EVM address._

**Risks:**
- Verify `getCanonicalMultiSendCallOnlyAddress` signature actually accepts `(chainId, version)` and that the new overload is backward-compatible
- Check that `chainId` is always defined at this call site — it is conditionally checked in the `if` guard but confirm no edge case where it could be undefined/empty string
- Confirm the function returns `undefined` or a safe fallback when no chain-specific address is registered, to avoid passing an invalid address downstream

### `packages/utils/src/services/contracts/__tests__/deployments.test.ts`

The test file for `deployments.test.ts` is heavily expanded: the `getCanonicalMultiSendCallOnlyAddress` describe block is rewritten with ~15 new parameterized cases covering fallback, no-canonical, warning logging, and null/undefined version handling; three new describe blocks are added for `eip155` and `zksync` deployment types using `getChainAgnosticAddress` and `resolveChainAgnosticContractAddresses`, with a dynamic `ZKSYNC_MULTISEND_CALL_ONLY_VERSIONS` filter and a guard test to prevent silent empty suites.

_These tests directly validate the fix that prevents `getChainAgnosticAddress` from returning a canonical/wrong-type address on chains where only a zkSync or eip155 deployment is registered, ensuring MultiSend transactions work correctly on zkSync Era and similar chains._

**Risks:**
- Chain ID constants (SOPHON='50104', ROPSTEN='3', FORWARD_ONLY_CHAIN='63', EIP155_ONLY_V130='16', CANONICAL_ONLY_CHAIN_ID='30') are hardcoded assumptions about safe-deployments data — if the upstream package changes its network registrations, tests will silently test wrong invariants or break unexpectedly
- ZKSYNC_MULTISEND_CALL_ONLY_VERSIONS is computed at module load time from live safe-deployments data; if the package is updated to remove zkSync entries, the guard test catches it, but the describe.each blocks would generate zero tests before the guard fires — verify test execution order guarantees the guard runs first
- The 'falls back from v1.5.0 to v1.4.1 canonical on zkSync (324)' test assumes zkSync chain 324 has v1.4.1 canonical but not v1.5.0 canonical — this is a fragile assumption tied to current safe-deployments state
- warnSpy uses mockImplementation to suppress console output in tests — verify afterEach mockRestore is sufficient and no test leaks spy state across describe blocks
- The cast `'1.2.0' as SafeState['version']` bypasses TypeScript type safety to test an out-of-range version — confirm the production code actually handles this gracefully rather than throwing

### `packages/utils/src/services/contracts/deployments.ts`

In `getChainAgnosticAddress`, a guard is added so that non-canonical `deploymentType` addresses (e.g. `'zksync'`) are only returned when explicitly registered in `networkAddresses[chainId]`; otherwise falls through to `getCanonicalOrFirstAddress`. `getCanonicalMultiSendCallOnlyAddress` gains a `chainId` parameter and now iterates `CANONICAL_FALLBACK_VERSIONS` (1.5.0→1.3.0) using `hasCanonicalDeployment` to find the newest canonical address actually registered on the chain, replacing a simple `deployment.deployments.canonical?.address` lookup.

_Prevents EraVM Safes on zkSync from receiving a canonical (EVM bytecode) MultiSendCallOnly address that was never deployed on the target chain, which would cause delegatecall failures._

**Risks:**
- All callers of `getCanonicalMultiSendCallOnlyAddress` must now pass `chainId` as the first argument — verify no call sites were missed or silently broken by the signature change
- The `??  []` fallback in `toNetworkAddressList` changes behavior for chains with no `networkAddresses` entry: previously `hasCanonicalDeployment` would have thrown/returned false on undefined; confirm `toNetworkAddressList([])` is safe and the semantics are correct
- CANONICAL_FALLBACK_VERSIONS is hardcoded to ['1.5.0','1.4.1','1.3.0']; adding a new Safe version requires manual update here
- The new guard in `getChainAgnosticAddress` only fires when `deploymentType !== 'canonical'`; confirm `'eip155'` is also handled correctly since it is neither `'canonical'` nor `'zksync'`
- semverSatisfies import must be present and correctly handles the `<=requested` range string — verify edge cases like exact version match and pre-release strings
- If `hasCanonicalDeployment` returns false for all candidates, a `console.warn` is emitted and `undefined` returned; callers must handle `undefined` gracefully to avoid downstream errors

---
_Generated by prbrief • model: claude-sonnet-4-6 • 6 LLM calls • $0.15 • 77.4s_

</details>

<sub>Generated locally by <a href="https://github.com/anthropics/claude-code">prbrief</a>. · `safe-global/safe-wallet-monorepo@9acf00025ee4`</sub>

**Comment by vmosharova** [comment]
Closing this PR as the issues was fixed here: https://github.com/safe-global/safe-wallet-monorepo/pull/7719
